src/hotspot/cpu/s390/templateTable_s390.cpp

```diff
@@ -2816,124 +2900,152 @@ void TemplateTable::jvmti_post_field_mod(Register cache,
   __ call_VM(noreg,
              CAST_FROM_FN_PTR(address, InterpreterRuntime::post_field_modification),
-             Z_ARG2, Z_ARG3, Z_ARG4);
-  __ get_cache_and_index_at_bcp(cache, index, 1);
+             obj, fieldEntry, value);
 
-  __ bind(L1);
+  // Reload field entry
+  __ load_field_entry(cache, index);
+
+  __ bind(dontPost);
   BLOCK_COMMENT("} jvmti_post_field_mod");
 }
 
 
 void TemplateTable::putfield_or_static(int byte_no, bool is_static, RewriteControl rc) {
   transition(vtos, vtos);
 
-  const Register cache         = Z_tmp_1;
-  const Register index         = Z_ARG5;
-  const Register obj           = Z_tmp_1;
+  const Register obj           = Z_ARG5;
   const Register off           = Z_tmp_2;
-  const Register flags         = Z_R1_scratch;
-  const Register br_tab        = Z_ARG5;
-  const Register bc            = Z_tmp_1;
+  const Register cache         = Z_ARG5;
+  const Register index         = Z_tmp_2;
+  const Register fieldAddr     = Z_tmp_2;      // contains obj and off combined. Could be any address register.
+  const Register flags         = Z_tmp_1;      // preserves flag value till the end, for volatility check
+  const Register br_tab        = Z_R1_scratch;
+  const Register tos_state     = Z_ARG4;
+  const Register bc_reg        = Z_tmp_2;
+  const Register patch_tmp     = Z_ARG4;
   const Register oopStore_tmp1 = Z_R1_scratch;
-  const Register oopStore_tmp2 = Z_ARG5;
+  const Register oopStore_tmp2 = Z_ARG5;       // tmp2 must be non-volatile reg
   const Register oopStore_tmp3 = Z_R0_scratch;
+#ifdef ASSERT
+  const Register br_tab_temp   = Z_R0_scratch; // for branch table verification code only
+#endif
 
-  resolve_cache_and_index(byte_no, cache, index, sizeof(u2));
+/*
+ *  Register usage and life range
+ *
+ *  cache, index          : short-lived. Their life ends after load_resolved_field_entry.
+ *  obj (overwrites cache): very short-lived, Combined with off immediately.
+ *  off (overwrites index): long-lived, Used in branch table entries.
+ *  flags                 : long-lived, Has to survive until the end to determine volatility.
+ *  br_tab                : short-lived, Only used to address branch table, and for verification in BTB_BEGIN macro.
+ *  tos_state             : short-live, Only used to index the branch table entry.
+ *  bc_reg                : short-lived, Used as work register in patch_bytecode.
+*/
+  resolve_cache_and_index_for_field(byte_no, cache, index);
   jvmti_post_field_mod(cache, index, is_static);
-  load_field_cp_cache_entry(obj, cache, index, off, flags, is_static);
-  // begin of life for:
-  //   obj, off   long life range
-  //   flags      short life range, up to branch into branch table
-  // end of life for:
-  //   cache, index
+  load_resolved_field_entry(obj, cache, tos_state, off, flags, is_static);
+
+  // Displacement is 0. No need to care about limited displacement range.
+  const Address field(fieldAddr);
+  __ lgr_if_needed(fieldAddr, off);
```

Comment:
Was there a problem with the range? Why did it work in the old implementation?

Reply:
The comment is related to the uses of field. Depending on the instruction, disc is limited to be uimm12 or simm20. But the Address object is with displacement == 0.

Reply:
I didn't get why the separate `agr` instructions are needed as part of this PR. Isn't that a separate issue?

Reply:
Martin, Are you referring to the fix for preserving `flag` register value till the end ?

Reply:
If you want that this case to be handled in separate PR, then I'll file a JBS issue.

Reply:
I need an explanation why this code was changed this way. If it is required for this PR, please explain in comments. If it's a separate issue, please handle separately.

Reply:
Oh, Issue was with the `flag` register value. We need to check the volatility at the end of `putfield_or_static` method. And previous implementation was not preserving the value till the end, why it was working then ? That I don't know. 

Current Implementation uses `z_tmp_1` for saving `flag` value till end. `obj` now points to `Z_ARG5` which is also pointed by `oopStore_tmp2`, and if we don't use `agr` then in case of `object/atos` , will run into `Multiple usecase for Z_R6` error from `do_oop_store()` method .

Reply:
Is the agr only used to save a register? That should be explained in the comments. I believe it's only needed for the object case, right?

Reply:
The prior implementation used two non-volatile registers. Due to register shortage, only one register can be used. Penalty: compute the sum of obj and off explicitly immediately after the obj pointer is popped from stack. In the non-static case, this can't be done upfront.